x-pack/plugins/security_solution/server/lib/detection_engine/rule_execution_log/saved_objects_adapter/saved_objects_adapter.ts

```diff
@@ -0,0 +1,187 @@
+import { IRuleStatusSOAttributes } from '../../rules/types';
+import {
+  RuleStatusSavedObjectsClient,
+  ruleStatusSavedObjectsClientFactory,
+} from './rule_status_saved_objects_client';
+import {
+  ExecutionMetric,
+  ExecutionMetricArgs,
+  FindBulkExecutionLogArgs,
+  FindExecutionLogArgs,
+  IRuleExecutionLogClient,
+  LegacyMetrics,
+  LogStatusChangeArgs,
+  UpdateExecutionLogArgs,
+} from '../types';
+
+// 1st is mutable status, followed by 5 most recent failures
+export const MAX_RULE_STATUSES = 6;
+
+const METRIC_FIELDS = {
+  [ExecutionMetric.executionGap]: 'gap',
+  [ExecutionMetric.searchDurationMax]: 'searchAfterTimeDurations',
+  [ExecutionMetric.indexingDurationMax]: 'bulkCreateTimeDurations',
+  [ExecutionMetric.indexingLookback]: 'lastLookBackDate',
+} as const;
+
+const getMetricField = <T extends ExecutionMetric>(metric: T) => METRIC_FIELDS[metric];
+
+export class SavedObjectsAdapter implements IRuleExecutionLogClient {
+  private ruleStatusClient: RuleStatusSavedObjectsClient;
+
+  constructor(savedObjectsClient: SavedObjectsClientContract) {
+    this.ruleStatusClient = ruleStatusSavedObjectsClientFactory(savedObjectsClient);
+  }
+
+  public find({ ruleId, logsCount = 1 }: FindExecutionLogArgs) {
+    return this.ruleStatusClient.find({
+      perPage: logsCount,
+      sortField: 'statusDate',
+      sortOrder: 'desc',
+      search: ruleId,
+      searchFields: ['alertId'],
+    });
+  }
+
+  public findBulk({ ruleIds, logsCount = 1 }: FindBulkExecutionLogArgs) {
+    return this.ruleStatusClient.findBulk(ruleIds, logsCount);
+  }
+
+  public async update({ id, attributes }: UpdateExecutionLogArgs) {
+    await this.ruleStatusClient.update(id, attributes);
+  }
+
+  public async delete(id: string) {
+    await this.ruleStatusClient.delete(id);
+  }
+
+  public async logExecutionMetric<T extends ExecutionMetric>({
+    ruleId,
+    metric,
+    value,
+  }: ExecutionMetricArgs<T>) {
+    const [currentStatus] = await this.getOrCreateRuleStatuses(ruleId);
+
+    await this.ruleStatusClient.update(currentStatus.id, {
+      ...currentStatus.attributes,
+      [getMetricField(metric)]: value,
+    });
+  }
+
+  private createNewRuleStatus = async (
+    ruleId: string
+  ): Promise<SavedObject<IRuleStatusSOAttributes>> => {
+    const now = new Date().toISOString();
+    return this.ruleStatusClient.create({
+      alertId: ruleId,
+      statusDate: now,
+      status: RuleExecutionStatus['going to run'],
+      lastFailureAt: null,
+      lastSuccessAt: null,
+      lastFailureMessage: null,
+      lastSuccessMessage: null,
+      gap: null,
+      bulkCreateTimeDurations: [],
+      searchAfterTimeDurations: [],
+      lastLookBackDate: null,
+    });
+  };
+
+  private getOrCreateRuleStatuses = async (
+    ruleId: string
+  ): Promise<Array<SavedObject<IRuleStatusSOAttributes>>> => {
+    const ruleStatuses = await this.find({
+      spaceId: '',
```

Comment:
I'd add a comment clarifying why space id is empty here.

Reply:
Added a comment